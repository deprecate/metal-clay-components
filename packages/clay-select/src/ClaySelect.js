import Component from 'metal-component';
import Soy from 'metal-soy';
import {Config} from 'metal-state';

import templates from './ClaySelect.soy.js';

/**
 * Metal Clay Select component.
 */
class ClaySelect extends Component {}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClaySelect.STATE = {
  /**
	 * The select disabled attribute.
	 * @instance
	 * @memberof ClaySelect
	 * @type {?boolean}
	 * @default false
	 */
  disabled: Config.bool().value(false),

  /**
	 * Id to be applied to the element.
	 * @instance
	 * @memberof ClaySelect
	 * @type {?string|undefined}
	 * @default undefined
	 */
  id: Config.string(),

  /**
	 * Label of the select element.
	 * @instance
	 * @memberof ClaySelect
	 * @type {?string|undefined}
	 * @default undefined
	 */
  label: Config.string(),

  /**
	 * Flag to indicate if the select element should be placed inline.
	 * @instance
	 * @memberof ClaySelect
	 * @type {?bool}
	 * @default false
	 */
  multiple: Config.bool().value(false),

  /**
	 * Name of the select element.
	 * @instance
	 * @memberof ClaySelect
	 * @type {?string|undefined}
	 * @default undefined
	 */
  name: Config.string(),

  /**
	 * Options of the select element.
	 * @instance
	 * @memberof ClaySelect
	 * @type {!array}
	 * @default undefined
	 */
  options: Config.arrayOf(
    Config.shapeOf({
      label: Config.string().required(),
      selected: Config.bool().value(false),
      value: Config.string().required(),
    })
  ).required(),

  /**
	 * Type of the select wrapper element.
	 * In order to make inline-group work properly you may place it inside
	 * a .form-inline.form-inline-autofit container.
	 * container.
	 * @instance
	 * @memberof ClaySelect
	 * @type {?string|undefined}
	 * @default group
	 */
  wrapperType: Config.oneOf(['inline-group', 'group']).value('group'),
};

Soy.register(ClaySelect, templates);

export {ClaySelect};
export default ClaySelect;
