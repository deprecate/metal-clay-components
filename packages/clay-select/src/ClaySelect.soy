{namespace ClaySelect}

/**
 * This renders the component's whole content.
 */
{template .render}
	{@param options: list<?>}
	{@param? disabled: bool}
	{@param? id: string}
	{@param? label: string}
	{@param? multiple: bool}
	{@param? name: string}
	{@param? wrapperType: string}

	{let $classes kind="text"}
		{if $wrapperType}
			form-{$wrapperType}
		{else}
			form-group
		{/if}
	{/let}

	<div class="{$classes}" id="{$id}">
		{if $label}
			<label>{$label}</label>
		{/if}

		{call .select}
			{param disabled: $disabled /}
			{param label: $label /}
			{param multiple: $multiple /}
			{param name: $name /}
			{param options: $options /}
		{/call}
	</div>
{/template}

/**
 * This renders the select tag.
 */
{template .select}
	{@param options: list<?>}
	{@param? disabled: bool}
	{@param? label: string}
	{@param? multiple: bool}
	{@param? name: string}

	{let $selectAttributes kind="attributes"}
		class="form-control"
		{if $disabled}
			disabled="disabled"
		{/if}
		{if $multiple}
			multiple="multiple"
		{/if}
		{if $name}
			name="{$name}"
		{/if}
		{if $label}
			aria-label="{$label}"
		{/if}
	{/let}

	<select {$selectAttributes}>
		{foreach $option in $options}
			{let $optionAttributes kind="attributes"}
				{if $option.selected}
					selected="selected"
				{/if}
				value="{$option.value}"
			{/let}

			<option {$optionAttributes}>{$option.label}</option>
		{/foreach}
	</select>
{/template}
