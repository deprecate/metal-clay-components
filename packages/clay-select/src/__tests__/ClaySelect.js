import ClaySelect from '../ClaySelect';

let select;

describe('ClaySelect', function() {
  afterEach(() => {
    if (select) {
      select.dispose();
    }
  });

  it('should generate a select element', function() {
    select = new ClaySelect({
      options: [
        {
          label: 'Option 1',
          value: '1',
        },
        {
          label: 'Option 2',
          value: '2',
        },
      ],
    });

    expect(select).toMatchSnapshot();
  });

  it('should generate a select element with label', function() {
    select = new ClaySelect({
      label: 'My Select',
      options: [
        {
          label: 'Option 1',
          value: '1',
        },
        {
          label: 'Option 2',
          value: '2',
        },
      ],
    });

    expect(select).toMatchSnapshot();
  });

  it('should generate a select element with name', function() {
    select = new ClaySelect({
      name: 'myselectname',
      options: [
        {
          label: 'Option 1',
          value: '1',
        },
        {
          label: 'Option 2',
          value: '2',
        },
      ],
    });

    expect(select).toMatchSnapshot();
  });

  it('should generate a select element with id', function() {
    select = new ClaySelect({
      id: 'myselectid',
      options: [
        {
          label: 'Option 1',
          value: '1',
        },
        {
          label: 'Option 2',
          value: '2',
        },
      ],
    });

    expect(select).toMatchSnapshot();
  });

  it('should generate an inline group select element', function() {
    select = new ClaySelect({
      label: 'My Select',
      options: [
        {
          label: 'Option 1',
          value: '1',
        },
        {
          label: 'Option 2',
          value: '2',
        },
      ],
      wrapperType: 'inline-group',
    });

    expect(select).toMatchSnapshot();
  });

  it('should generate a select element with preselected option', function() {
    select = new ClaySelect({
      options: [
        {
          label: 'Option 1',
          value: '1',
        },
        {
          label: 'Option 2',
          selected: true,
          value: '2',
        },
      ],
    });

    expect(select).toMatchSnapshot();
  });

  it('should generate a disabled select element', function() {
    select = new ClaySelect({
      disabled: true,
      options: [
        {
          label: 'Option 1',
          value: '1',
        },
        {
          label: 'Option 2',
          value: '2',
        },
      ],
    });

    expect(select).toMatchSnapshot();
  });

  it('should generate a multiple select element', function() {
    select = new ClaySelect({
      multiple: true,
      options: [
        {
          label: 'Option 1',
          value: '1',
        },
        {
          label: 'Option 2',
          value: '2',
        },
      ],
    });

    expect(select).toMatchSnapshot();
  });

  it('should generate an inline group multiple select element', function() {
    select = new ClaySelect({
      multiple: true,
      options: [
        {
          label: 'Option 1',
          value: '1',
        },
        {
          label: 'Option 2',
          value: '2',
        },
      ],
      wrapperType: 'inline-group',
    });

    expect(select).toMatchSnapshot();
  });
});
