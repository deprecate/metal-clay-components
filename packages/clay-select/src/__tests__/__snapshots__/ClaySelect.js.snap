// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClaySelect should generate a disabled select element 1`] = `
<div class="form-group">
  <select class="form-control" disabled="disabled">
    <option value="1">Option 1</option>
    <option value="2">Option 2</option>
  </select>
</div>
`;

exports[`ClaySelect should generate a multiple select element 1`] = `
<div class="form-group">
  <select class="form-control" multiple="multiple">
    <option value="1">Option 1</option>
    <option value="2">Option 2</option>
  </select>
</div>
`;

exports[`ClaySelect should generate a select element 1`] = `
<div class="form-group">
  <select class="form-control">
    <option value="1">Option 1</option>
    <option value="2">Option 2</option>
  </select>
</div>
`;

exports[`ClaySelect should generate a select element with id 1`] = `
<div class="form-group" id="myselectid">
  <select class="form-control">
    <option value="1">Option 1</option>
    <option value="2">Option 2</option>
  </select>
</div>
`;

exports[`ClaySelect should generate a select element with label 1`] = `
<div class="form-group">
  <label>My Select</label>
  <select class="form-control" aria-label="My Select">
    <option value="1">Option 1</option>
    <option value="2">Option 2</option>
  </select>
</div>
`;

exports[`ClaySelect should generate a select element with name 1`] = `
<div class="form-group">
  <select class="form-control" name="myselectname">
    <option value="1">Option 1</option>
    <option value="2">Option 2</option>
  </select>
</div>
`;

exports[`ClaySelect should generate a select element with preselected option 1`] = `
<div class="form-group">
  <select class="form-control">
    <option value="1">Option 1</option>
    <option selected="selected" value="2">Option 2</option>
  </select>
</div>
`;

exports[`ClaySelect should generate an inline group multiple select element 1`] = `
<div class="form-inline-group">
  <select class="form-control" multiple="multiple">
    <option value="1">Option 1</option>
    <option value="2">Option 2</option>
  </select>
</div>
`;

exports[`ClaySelect should generate an inline group select element 1`] = `
<div class="form-inline-group">
  <label>My Select</label>
  <select class="form-control" aria-label="My Select">
    <option value="1">Option 1</option>
    <option value="2">Option 2</option>
  </select>
</div>
`;
